pkg/gci/mod.go

```diff
@@ -0,0 +1,109 @@
+package gci
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/mod/modfile"
+)
+
+// moduleResolver looksup the module path for a given (Go) file.
+// To improve performance, the file paths and module paths are
+// cached.
+//
+// Given the following directory structure:
+//
+//	/path/to/example
+//	+-- go.mod               (module example)
+//	+-- cmd/sample/main.go   (package main, imports example/util)
+//	+-- util/util.go         (package util)
+//
+// After looking up main.go and util.go, the internal cache will contain:
+//
+//	"/path/to/foobar/": "example"
+//
+// For more complex module structures (i.e. sub-modules), the cache
+// might look like this:
+//
+//	"/path/to/example/":            "example"
+//	"/path/to/example/cmd/sample/": "go.example.com/historic/path"
+//
+// When matching files against this cache, the resolver will select the
+// entry with the most specific path (so that, in this example, the file
+// cmd/sample/main.go will resolve to go.example.com/historic/path).
+type moduleResolver map[string]string
+
+var modCache = make(moduleResolver)
+
+// ClearModCache will reset the internal module cache used by ProcessFile
+// and WalkDir. When invoking either of those functions with an empty
+// FlagSet.LocalFlag, we will try to resolve a matching Go module path.
+//
+// You should call this method when you're done with processing files.
+func ClearModCache() {
+	modCache = make(moduleResolver)
+}
+
+func (m moduleResolver) Lookup(file string) (string, error) {
+	var bestMatch string
+	for path := range m {
+		if strings.HasPrefix(file, path) && len(path) > len(bestMatch) {
+			bestMatch = path
+		}
```

Comment:
Hi, I very much like this PR, but was a bit confused here. The cache `m` seems to be indexed by absolute paths, so maybe we should calculate the absolute path of `file` before searching it?